R/fdbk_asdataframe.R

```diff
@@ -1156,6 +1156,58 @@ comparableRows <- function(DT,splitCol,splitVal,compareBy){
 ########################################################################################################################
 
 
+#' Bin a data.table column around user defined levels and replace it with the levels value.
+#' 
+#' @description Other way to perform a binning like in function fdbk_dt_binning but by defining levels around which to bin instead of the bins limits. 
+#' The limits of the bins will be calculated by taking the mean between neighbouring levels. The two functions differ in the sense that 
+#' fdbk_dt_binning allow to have gaps between the bins, whereas the bins will be continuous in fdbk_dt_binning_level. This function allows to have non-equally spaced 
+#' levels without gaps between the bins, so that the level is not always at the center of the bin. 
+#' @param  DT data.table
+#' @param  varToBin variable that should be binned (and will be replaced by the binned version)
+#' @param  levels number/vector of levels on which the bins will be defined
+#' @return data.table with varToBin replaced by factorized mid-bin values  (NA if variable falls in none of the bins)
+#
+#' @author Josue <josue.gehring@@gmail.com>
+#'
+#' @seealso \code{\link{cut}}
+#'
+#' @examples 
+#' #plot scores accross binned levels
+#' require(ggplot2)
+#' fnames       = "~/examplesRfdbk/icon/temp/verTEMP.2014120112"
+#' cond        = list(obs="!is.na(obs)",varno="varno%in%c(2,3,4,29)")
+#' columnnames = c("obs","veri_data","varno","state","level")
+#' DT          = fdbk_dt_multi_large(fnames,cond,columnnames,1)
+#' levels = c(100000,92500,85000,70000,60000,50000,40000,30000)
+#' DT = fdbk_dt_binning_level(DT,"level",levels)
+#' DT$varno    = varno_to_name(DT$varno)
+#' strat       = c("varno","level")
+#' scores      = fdbk_dt_verif_continuous(DT,strat)
+#' setkey(scores,scorename,varno,level)
+#' scores      = scores[!is.na(scores),]
+#' p =  ggplot(scores,aes(x=scores,y=level,group=interaction(varno,scorename)))+
+#'  geom_path() + facet_wrap(~varno~scorename,scales="free_x",ncol = 6)+
+#'  theme_bw()+theme(axis.text.x  = element_text(angle=70,hjust = 1))+scale_y_reverse()
+#' p
+fdbk_dt_binning_level <- function(DT,varToBin="level",levels){
+  bins = getVarToBin(DT,varToBin) 
+  dp = diff(levels)/2
+  binUpper = c(levels[1]-dp[1],levels[1:length(levels)-1]+dp) 
+  binLower = c(levels[1:length(levels)-1]+dp,levels[length(levels)]+dp[length(dp)])
+  
+  for (i in 1:length(binLower)){
+    bins[DT[,varToBin,with=F]>=binLower[i] & DT[,varToBin,with=F]<binUpper[i] ] = levels[i]
```

Comment:
I am wondering whether an easier way to achieve this would be:
dp <- diff(levels)/2
levelBnds <- c(levels[1], levels) + c(-dp[1], dp, dp[length(dp)])
bins <- levels[findInterval(DT[,varToBin,with=F], levelBnds)]
DT[,varToBin] <- bins